cluster-autoscaler/cloudprovider/gce/mig_instance_templates_provider.go

```diff
@@ -52,14 +52,13 @@ func NewCachingMigInstanceTemplatesProvider(cache *GceCache, gceClient Autoscali
 // GetMigInstanceTemplate returns instance template for MIG with given ref
 func (p *CachingMigInstanceTemplatesProvider) GetMigInstanceTemplate(migRef GceRef) (*gce.InstanceTemplate, error) {
 	p.mutex.Lock()
-	defer p.mutex.Unlock()
-
 	if !p.lastRefresh.Add(migInstanceCacheRefreshInterval).After(time.Now()) {
 		p.cache.InvalidateAllMigInstanceTemplates()
 		p.lastRefresh = time.Now()
 	}
 
 	instanceTemplate, found := p.cache.GetMigInstanceTemplate(migRef)
+	p.mutex.Unlock()
```

Comment:
releasing and acquire locks multiple times in a function can cause bugs where a refactor can break things. Consider making moving each snippet into new funcs where each func begins with lock and defer unlock